File: apps/quickink/android/app/src/main/java/app/quickink/mobile/features/scan/cardcapture/CardGuideOverlay.kt
/*
 * CardGuideOverlay.kt
 *
 * The translucent business-card-shaped guide drawn over the
 * camera preview. Three responsibilities:
 *
 *   1. Frame the 1.586:1 / 70%-of-preview-width target so the
 *      user knows where to hold the card.
 *   2. Tint the frame neutral / yellow / green based on
 *      detection state.
 *   3. Render the "Hold steady…" / "Position card inside the
 *      frame" hint text below the guide.
 *
 * Pure Compose — no Android-View interop. The parent surface
 * composes this on top of the PreviewView and feeds it the
 * current [OverlayState] from the analyzer.
 */

package app.quickink.mobile.features.scan.cardcapture

import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.offset
import androidx.compose.material3.Text
import androidx.compose.ui.Alignment
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.geometry.RoundRect
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.PathOperation
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.unit.dp
import app.quickink.mobile.ui.theme.LocalQuickInkTypography

/**
 * What the overlay is being told to show. Mirrors the spec's
 * neutral / yellow / green visual states 1:1.
 */
enum class OverlayState {
    /** No valid quad anywhere. Frame is neutral; hint blank. */
    Neutral,
    /** A quad is being seen but failed at least one gate. Frame yellow; hint "Hold the card flat". */
    Partial,
    /** A valid quad — within IoU, aspect, etc. Frame green; hint "Hold steady…". */
    Valid,
    /** No detection for ≥8 s. Frame neutral; hint "Position card inside the frame". */
    Idle,
}

/**
 * Card-frame dimensions in canvas pixels. The same arithmetic
 * powers both the drawn rect and the analyzer-frame
 * [GuideRect] — the surface re-derives the analyzer-space rect
 * by scaling these numbers by analyzer-width / canvas-width.
 */
data class GuideMetrics(
    val rect: Rect,
    val cornerRadiusPx: Float,
) {
    companion object {
        /** ISO 7810 ID-1 aspect — same as the spec's 1.586:1. */
        const val CARD_ASPECT_RATIO = 1.586f
        /** Guide width as a fraction of the preview width. */
        const val GUIDE_WIDTH_FRACTION = 0.70f

        fun compute(canvasWidth: Float, canvasHeight: Float): GuideMetrics {
            val targetW = canvasWidth * GUIDE_WIDTH_FRACTION
            val targetH = targetW / CARD_ASPECT_RATIO
            val cx = canvasWidth * 0.5f
            // 45% rather than 50% — gives the hint text room
            // between the guide and the shutter row below.
            val cy = canvasHeight * 0.45f
            val left = cx - targetW / 2f
            val top  = cy - targetH / 2f
            return GuideMetrics(
                rect = Rect(left, top, left + targetW, top + targetH),
                cornerRadiusPx = targetW * 0.04f,
            )
        }
    }
}

@Composable
fun CardGuideOverlay(
    state: OverlayState,
    onMetricsKnown: (GuideMetrics) -> Unit,
    modifier: Modifier = Modifier,
) {
    val frameColor by animateColorAsState(
        targetValue = when (state) {
            OverlayState.Valid   -> Color(0xFF34D399) // green
            OverlayState.Partial -> Color(0xFFFACC15) // yellow
            OverlayState.Neutral,
            OverlayState.Idle    -> Color.White.copy(alpha = 0.65f)
        },
        animationSpec = tween(durationMillis = 150),
        label = "card-overlay-frame-color",
    )

    val hintText = when (state) {
        OverlayState.Valid   -> "Hold steady…"
        OverlayState.Partial -> "Hold the card flat"
        OverlayState.Idle    -> "Position card inside the frame"
        OverlayState.Neutral -> ""
    }

    val type = LocalQuickInkTypography.current
    val density = LocalDensity.current

    BoxWithConstraints(modifier = modifier.fillMaxSize()) {
        val canvasWidthPx  = with(density) { maxWidth.toPx() }
        val canvasHeightPx = with(density) { maxHeight.toPx() }
        val metrics = GuideMetrics.compute(canvasWidthPx, canvasHeightPx)

        // Publish metrics back to the surface so it can compute
        // the analyzer-space GuideRect for the detector. Fires
        // every recomposition that changes constraints; safe to
        // pass the same struct repeatedly since the surface
        // stores it in `remember { mutableStateOf(...) }`.
        onMetricsKnown(metrics)

        Canvas(modifier = Modifier.fillMaxSize()) {
            // Punched-hole dim — full-screen scrim with the card
            // rect cut out. Path difference so the alpha doesn't
            // multiply against the preview.
            val scrim = Path().apply {
                addRect(Rect(Offset.Zero, Size(size.width, size.height)))
            }
            val cardHole = Path().apply {
                addRoundRect(
                    RoundRect(
                        rect = metrics.rect,
                        cornerRadius = CornerRadius(metrics.cornerRadiusPx, metrics.cornerRadiusPx),
                    )
                )
            }
            val masked = Path().apply {
                op(scrim, cardHole, PathOperation.Difference)
            }
            drawPath(path = masked, color = Color(0xCC0F0E0D))

            // Card-shaped frame stroke.
            drawRoundRect(
                color  = frameColor,
                topLeft = Offset(metrics.rect.left, metrics.rect.top),
                size   = Size(metrics.rect.width, metrics.rect.height),
                cornerRadius = CornerRadius(metrics.cornerRadiusPx, metrics.cornerRadiusPx),
                style  = Stroke(width = with(density) { 3.dp.toPx() }),
            )
        }

        // Hint text — horizontally centered inside the overlay,
        // offset vertically so it lands just below the guide
        // rect's bottom edge.
        if (hintText.isNotEmpty()) {
            val hintTopDp = with(density) { (metrics.rect.bottom + 24f).toDp() }
            Box(
                modifier = Modifier
                    .offset(x = 0.dp, y = hintTopDp)
                    .fillMaxWidth(),
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    text  = hintText,
                    style = type.label,
                    color = Color.White.copy(alpha = 0.85f),
                )
            }
        }
    }
}

File: apps/quickink/ios/QuickInk/Scan/CardCapture/CardGuideOverlay.swift
/*
 * CardGuideOverlay.swift
 *
 * Translucent business-card-shaped guide drawn over the
 * AVCaptureSession preview. Three responsibilities (same as
 * Android's `CardGuideOverlay.kt`):
 *
 *   1. Frame the 1.586:1 / 70%-of-preview-width target so the
 *      user knows where to hold the card.
 *   2. Tint the frame neutral / yellow / green based on
 *      detection state.
 *   3. Render the "Hold steady…" / "Position card inside the
 *      frame" hint text below the guide.
 *
 * SwiftUI-pure — no UIView interop. The parent surface stacks
 * this on top of a video preview layer via ZStack.
 */

import SwiftUI

public enum OverlayState: Sendable {
    /// No valid quad anywhere. Neutral tint; hint blank.
    case neutral
    /// A quad is being seen but failed at least one gate. Yellow tint.
    case partial
    /// A valid quad — passed all gates. Green tint.
    case valid
    /// No detection for ≥8 s. Neutral tint; "Position card" hint.
    case idle
}

public struct GuideMetrics: Equatable, Sendable {
    public let rect: CGRect
    public let cornerRadius: CGFloat
    /// The canvas size this metrics block was computed against,
    /// in points. Carried through to the detector + post-
    /// processor so they can apply the same resizeAspectFill
    /// center-crop math the preview layer uses on-screen.
    public let canvasSize: CGSize

    /// ISO 7810 ID-1 aspect — same number as the spec's 1.586:1.
    public static let cardAspectRatio: CGFloat = 1.586
    /// Target guide width as a fraction of the preview width.
    public static let guideWidthFraction: CGFloat = 0.70

    public static func compute(canvasSize: CGSize) -> GuideMetrics {
        let targetW = canvasSize.width * guideWidthFraction
        let targetH = targetW / cardAspectRatio
        let cx = canvasSize.width * 0.5
        // 45% rather than 50% — leaves room for the hint text
        // between the guide and the shutter row.
        let cy = canvasSize.height * 0.45
        let rect = CGRect(
            x: cx - targetW / 2,
            y: cy - targetH / 2,
            width: targetW,
            height: targetH,
        )
        return GuideMetrics(
            rect: rect,
            cornerRadius: targetW * 0.04,
            canvasSize: canvasSize,
        )
    }
}

public struct CardGuideOverlay: View {

    public let state: OverlayState
    public let onMetricsKnown: (GuideMetrics) -> Void

    public init(state: OverlayState, onMetricsKnown: @escaping (GuideMetrics) -> Void) {
        self.state = state
        self.onMetricsKnown = onMetricsKnown
    }

    public var body: some View {
        GeometryReader { geo in
            let metrics = GuideMetrics.compute(canvasSize: geo.size)
            ZStack {
                // Punched-hole dim — scrim everywhere except the
                // card rect. SwiftUI achieves the cutout via an
                // even-odd-filled path that includes both the
                // full canvas and the rounded card rect; only
                // the difference renders as filled.
                Path { path in
                    path.addRect(CGRect(origin: .zero, size: geo.size))
                    path.addRoundedRect(
                        in: metrics.rect,
                        cornerSize: CGSize(width: metrics.cornerRadius, height: metrics.cornerRadius),
                    )
                }
                .fill(Color(red: 15/255, green: 14/255, blue: 13/255).opacity(0.80), style: FillStyle(eoFill: true))

                // Card-frame stroke.
                RoundedRectangle(cornerRadius: metrics.cornerRadius, style: .continuous)
                    .stroke(frameColor, lineWidth: 3)
                    .frame(width: metrics.rect.width, height: metrics.rect.height)
                    .position(x: metrics.rect.midX, y: metrics.rect.midY)
                    .animation(.easeInOut(duration: 0.15), value: state)

                // Hint text — anchored below the guide rect.
                if !hintText.isEmpty {
                    Text(hintText)
                        .font(QuickInkText.label)
                        .foregroundStyle(Color.white.opacity(0.85))
                        .position(
                            x: geo.size.width * 0.5,
                            y: metrics.rect.maxY + 32,
                        )
                }
            }
            .onAppear { onMetricsKnown(metrics) }
            .onChange(of: metrics) { newMetrics in onMetricsKnown(newMetrics) }
        }
    }

    private var frameColor: Color {
        switch state {
        case .valid:                return Color(red: 0x34/255, green: 0xD3/255, blue: 0x99/255)
        case .partial:              return Color(red: 0xFA/255, green: 0xCC/255, blue: 0x15/255)
        case .neutral, .idle:       return Color.white.opacity(0.65)
        }
    }

    private var hintText: String {
        switch state {
        case .valid:   return "Hold steady…"
        case .partial: return "Hold the card flat"
        case .idle:    return "Position card inside the frame"
        case .neutral: return ""
        }
    }
}
